Support exclusion bounds in power distributor

shsms opened the pull request:
To be merged after #537 

Closes #152

shsms committed:
Add a `_get_bounds` method in PowerDistributingActor

This method will replace the `_get_{upper,lower}_bound` methods in
subsequent commits.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Respect exclusion bounds in `PowerDistributingActor._check_request`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove unused `to_protobuf` method in power distribution tests

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Support exclusion bounds in power distribution tests

This just updates the tests to use data structures that support
exclusion bounds, so that once support is implemented in the actor,
it would be easy to add incremental tests.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Refactor to reduce the amount of calls to the component data cache

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update distribution algorithm to support exclusion bounds

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
Still need to add tests.

shsms committed:
Extract reusable `create_components` method outside test class

This allows us to not pack all tests in the same class, but group them
separately in smaller units.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Test exclusion bounds in power distribution algorithm

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update RELEASE_NOTES.md

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

llucax commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -472,19 +520,27 @@ def _check_request(self, request: Request) -> Optional[Result]:
-                if request.power > bound:
-                    return OutOfBound(request=request, bound=bound)
+        bounds = self._get_bounds(pairs_data)
+        if request.adjust_power:
+            # Automatic power adjustments can only bring down the requested power down
+            # to the inclusion bounds.
+            #
+            # If the requested power is in the exclusion bounds, it is NOT possible to
+            # increase it so that it is outside the exclusion bounds.
```
Why is it not OK to increase power but it's OK to decrease power when adjusting was requested?

shsms replied:
This has been our practice I think.  Going towards zero can be automatic/not away from.  Increasing power has to be a considered decision, not automatic, because for example the minimum might be too high for sustained maintenance, and could lead to peaks or unexpected soc drops, etc.

llucax commented on tests/actor/test_power_distributing.py:
```diff
@@ -396,7 +397,7 @@ async def test_power_distributor_invalid_battery_id(
         assert isinstance(result, Error)
         assert result.request == request
-        err_msg = re.search(r"^No battery 100, available batteries:", result.msg)
+        err_msg = re.search(r"No battery 100, available batteries:", result.msg)
```
Why is the `^` removed? It doesn't look like the message changed.

```suggestion
        err_msg = re.search(r"^No battery 100, available batteries:", result.msg)
```

shsms replied:
If you look at the corresponding commit,  https://github.com/frequenz-floss/frequenz-sdk-python/pull/562/commits/de1331d6bbc6adbca8dafd1198534456e087c91d, `check_request` call is moved downwards,  and so this error is caught and raised by a different `send_result` which wraps an error,  so the string is wrapped by quotes.

llucax replied:
Ah, ok, It was hidden in the diff, expanding it now I can see it, thanks for the clarification!

llucax commented on src/frequenz/sdk/power/_distribution_algorithm.py:
```diff
@@ -336,57 +346,98 @@ def _distribute_power(
 
         distribution: Dict[int, float] = {}
-
+        print(f"{power_w=}")
```
Leftover `print`

```suggestion
```

llucax commented on src/frequenz/sdk/power/_distribution_algorithm.py:
```diff
@@ -336,57 +346,98 @@ def _distribute_power(
+            distribution[inverter.component_id] = excl_bound
+
+        for inverter_id, deficit in deficits.items():
+            while not math.isclose(deficit, 0.0, abs_tol=1e-6) and deficit < 0.0:
```
You could use `._internal._math.is_close_to_zero()` here.
```suggestion
            while not is_close_to_zero(deficit) and deficit < 0.0:
```

llucax commented on src/frequenz/sdk/power/_distribution_algorithm.py:
```diff
@@ -336,57 +346,98 @@ def _distribute_power(
+        for inverter_id, deficit in deficits.items():
+            while not math.isclose(deficit, 0.0, abs_tol=1e-6) and deficit < 0.0:
+                take_from = max(excess_reserved.items(), key=lambda item: item[1])
+                if math.isclose(take_from[1], 0.0, abs_tol=1e-6) or take_from[1] < 0.0:
```
```suggestion
                if is_close_to_zero(take_from[1]) or take_from[1] < 0.0:
```

llucax commented on src/frequenz/sdk/power/_distribution_algorithm.py:
```diff
@@ -336,57 +346,98 @@ def _distribute_power(
+                if math.isclose(take_from[1], 0.0, abs_tol=1e-6) or take_from[1] < 0.0:
+                    break
+                if take_from[1] >= -deficit or math.isclose(
+                    take_from[1], -deficit, abs_tol=1e-6
```
I `deficit` supposed to be close to zero? `abs_tol` seems to be [mainly useful for comparing to near-zero values](https://docs.python.org/3/library/math.html#math.isclose) AFAIK, not sure if it is generally OK to use it for other values or not. Maybe @matthias-wende-frequenz can speak his mathematician mind...

shsms replied:
This is a `>=`.  But the `=` part is not exact, value is sometimes slightly on the other side.  hence the `isclose`.

llucax replied:
Yeah, what I'm asking if it is OK is specifically the `abs_tol` argument.

matthias-wende-frequenz replied:
I don't understand the full algorithm but the check is fine. It's just about having a higher certainty the the `deficit` is close the excess_reserved under consideration (the `take_from[1]`).

daniel-zullo-frequenz replied:
I think this should use the relative tolerance instead as `take_from[1]` and `deficit` are already compared to zero above where the absolute tolerance makes more sense

llucax replied:
The relative tolerance is defined by default (`rel_tol=1e-09`), this is why I asked only about removing `tol_abs`.

llucax reviewed (commented):
Except for the `abs_tol` for `deficit` (which I'm not sure it is an issue or not), nothing major from my side. I didn't dig very deep into the distribution algorithm itself. The rest can be easily applied by accepting the suggestions.

daniel-zullo-frequenz commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -472,19 +520,27 @@ def _check_request(self, request: Request) -> Optional[Result]:
+            if (
+                request.power > bounds.exclusion_lower
+                and request.power < bounds.exclusion_upper
+            ):
```
python:tm: 
```suggestion
            if bounds.exclusion_lower < request.power < bounds.exclusion_upper:
```

daniel-zullo-frequenz commented on src/frequenz/sdk/power/_distribution_algorithm.py:
```diff
@@ -336,57 +346,98 @@ def _distribute_power(
+                if left_over > -deficit:
+                    distributed_power += deficit
+                    deficit = 0.0
+                    deficits[inverter_id] = 0.0
```
I think `deficits` does not need to be updated in this for-loop

shsms replied:
true,  only the `distributed_power` needs to be updated.  I guess I added the other two just to be able to properly reason about what was going on as I was developing,  but they don't have to be there.

llucax replied:
Maybe add a comment about it in the refactor PR if it would make reasoning about it easier.

daniel-zullo-frequenz commented on src/frequenz/sdk/power/_distribution_algorithm.py:
```diff
@@ -336,57 +346,98 @@ def _distribute_power(
+                elif left_over > 0.0:
+                    deficit += left_over
+                    distributed_power += left_over
+                    deficits[inverter_id] = deficit
```
Updating `deficits` might be redundant here as it will no longer be used after the for-loop

daniel-zullo-frequenz commented on src/frequenz/sdk/power/_distribution_algorithm.py:
```diff
@@ -336,57 +346,98 @@ def _distribute_power(
+            distribution[inverter_id] += excess
+            distributed_power += excess
+
+        for inverter_id, deficit in deficits.items():
```
This duplicated for-loop is not needed as its content could be added in the first loop for deficits, right after the end of the `while`-loop

llucax committed:
Remove leftover `print()`

Signed-off-by: Leandro Lucarella <luca@llucax.com>

llucax committed:
Use `is_close_to_zero()` instead of `math.isclose(..., abs_tol=...)`

We already have an utility function to check if a `float` is close
to zero, so we can use that instead.

Signed-off-by: Leandro Lucarella <luca@llucax.com>

llucax committed:
Simplify `a < b < c` comparison

Co-authored-by: daniel-zullo-frequenz <120166726+daniel-zullo-frequenz@users.noreply.github.com>
Signed-off-by: Leandro Lucarella <luca@llucax.com>

llucax commented:
* https://github.com/frequenz-floss/frequenz-sdk-python/issues/572
